uwrt_mars_rover_vision/src/aruco_target_tracker.cpp

```diff
@@ -0,0 +1,146 @@
+                                rot_matrix.at<double>(2, 0), rot_matrix.at<double>(2, 1), rot_matrix.at<double>(2, 2)};
+  tf2::Vector3 tf2_tvec{ tvec[0], tvec[1], tvec[2] };
+  // Create a transform and convert to a Pose
+  tf2::Transform tf2_transform {tf2_rot_matrix, tf2_tvec };
+  tf2::toMsg(tf2_transform, pose_msg);
+}
+
+void TargetTracker::imageCallback(const sensor_msgs::msg::Image::ConstSharedPtr& image_msg) 
+{
+  cv_bridge::CvImagePtr cv_ptr;
+  try {
+    cv_ptr = cv_bridge::toCvCopy(image_msg, image_msg->encoding);
+  } catch (cv_bridge::Exception& e) {
+    RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
+    return;
+  }
+  // receive a bgra8 encoded image, must convert to bgr
+  cv::Mat image_rgb;
+  cv::cvtColor(cv_ptr->image, image_rgb, cv::COLOR_BGRA2BGR);
+
+  // detect ARUCO markers in the image
+  std::vector<int> marker_ids;
+  std::vector<std::vector<cv::Point2f>> marker_corners, rejection_candidates;
+  cv::aruco::detectMarkers(image_rgb, dictionary_, marker_corners, marker_ids, params_, rejection_candidates);
+  int num_markers_detected = marker_corners.size();
+  
+  // identify visible aruco marker poses
+  if (num_markers_detected > 0) 
+  {
+    // Calculate pose for each marker
+    for (int i = 0; i < num_markers_detected; i++) {
+      cv::solvePnP(obj_points_, marker_corners.at(i), intrinsic_calib_matrix_, dist_coefficients_, rvecs_.at(i),
+                   tvecs_.at(i));
+      // send a pose message stamped of the aruco tag pose
+      geometry_msgs::msg::PoseStamped aruco_pose_msg;
+      aruco_pose_msg.header.frame_id = std::to_string(marker_ids[i]);
+      toPoseMsg(rvecs_[i], tvecs_[i], aruco_pose_msg.pose);
+      aruco_pose_pub_->publish(aruco_pose_msg);
+      if (display_marker_pose_)
+      {
+        // draw the detected poses with coordinate axis of 0.2m in length
+        cv::drawFrameAxes(image_rgb, intrinsic_calib_matrix_, dist_coefficients_, rvecs_.at(i), tvecs_.at(i), 0.2);
+      }
+    }
+    RCLCPP_DEBUG_STREAM(this->get_logger(), "ARUCO at tvec: " << tvecs_.at(0) << " | rvec: " << rvecs_.at(0));
+    if (display_marker_pose_)
+    {
+      cv::aruco::drawDetectedMarkers(image_rgb, marker_corners, marker_ids);
+    }
+  } 
+
+  if (display_marker_pose_)
+  {
+    cv::imshow("Image window", image_rgb);
+    cv::waitKey(5);
+  }
+}
+
+void TargetTracker::camInfoCallback(const sensor_msgs::msg::CameraInfo::SharedPtr info) 
+{
+  // hacky work around to only read cam info once
+  if (intrinsic_calib_matrix_.size().height == 0)
+  {
+  RCLCPP_DEBUG_STREAM(this->get_logger(), info->distortion_model);
+  RCLCPP_DEBUG_STREAM(this->get_logger(), "fx: " << info->k[0] << " | cx: " << info->k[2] << " | fy: " << info->k[4] << " | cy: " << info->k[5]);
+  int param_num = 0; 
+  for (const double &disto_param: info->d)
+  {
+      dist_coefficients_.at<double>(0, param_num) = disto_param;
+      param_num++;
+  }
+  intrinsic_calib_matrix_ = (cv::Mat_<float>(3,3) <<
```

Comment:
what are these magic numbers?

Reply:
this is just making a 3x3 matrix :) think they arent magic numbers for that reason (ie they explicitly create that)

Reply:
o.O